include/RAJA/pattern/kernel/ArgHelper.hpp

```diff
@@ -0,0 +1,293 @@
+    -> camp::list<OffSet<id>>
+  {
+    return camp::list<OffSet<id>>{};
+  }
+};
+
+//Converts OffSetList<1,2,3> -> list<OffSet<0>, OffSet<1>, OffSet<2>>
+template<camp::idx_t head, camp::idx_t... tail>
+struct listMaker<OffSetList<head,tail...>>
+{
+  RAJA_HOST_DEVICE
+  static auto genList()
+    -> decltype(catList<camp::list<OffSet<head>>,
+            decltype(listMaker<OffSetList<tail...>>::genList())>::makeList(
+            camp::list<OffSet<head>>{},
+            listMaker<OffSetList<tail...>>::genList()))
+  {
+
+    return catList<camp::list<OffSet<head>>,
+            decltype(listMaker<OffSetList<tail...>>::genList())>::makeList(
+            camp::list<OffSet<head>>{},
+            listMaker<OffSetList<tail...>>::genList());
+  }
+};
+
+//Converts ParamList<1,2,3> -> list<Param<0>, Param<1>, Param<2>>
+template<camp::idx_t head, camp::idx_t... tail>
+struct listMaker<ParamList<head,tail...>>
+{
+  RAJA_HOST_DEVICE
+  static auto genList()
+    -> decltype(catList<camp::list<Param<head>>,
+            decltype(listMaker<ParamList<tail...>>::genList())>::makeList(
+            camp::list<Param<head>>{},
+            listMaker<ParamList<tail...>>::genList()))
+  {
+
+    return catList<camp::list<Param<head>>,
+            decltype(listMaker<ParamList<tail...>>::genList())>::makeList(
+            camp::list<Param<head>>{},
+            listMaker<ParamList<tail...>>::genList());
+  }
+};
+
+
+//Converts Param<id> -> list<Param<id>>
+template<camp::idx_t id>
+struct listMaker<Param<id>>
+{
+  RAJA_HOST_DEVICE
+  static auto genList()
+    -> camp::list<Param<id>>
+  {
+
+    return camp::list<Param<id>>{};
+  }
+};
+
+//Helper structs to parse through template arguments
+//Takes a pre-processing step convert SegList, ParamList
+//into a list of Seg and Param's
+template<typename List>
+struct parser{};
+
+template<>
+struct parser<camp::list<>>
+{
+  RAJA_HOST_DEVICE
+  static auto checkArgs()
+    -> camp::list<>
+  {
+    return camp::list<> {};
+  }
+};
+
+template <typename Head, typename... Tail>
+struct parser<camp::list<Head, Tail...>>
+{
+
+  RAJA_HOST_DEVICE
+  static auto checkArgs()
+    -> decltype(catList<decltype(listMaker<Head>::genList()),
+                decltype(parser<camp::list<Tail...> >::checkArgs())>
+                ::makeList(listMaker<Head>::genList(), parser<camp::list<Tail...> >::checkArgs()))
+  {
+
+    return catList<decltype(listMaker<Head>::genList()),
+                   decltype(parser<camp::list<Tail...> >::checkArgs())>
+      ::makeList(listMaker<Head>::genList(), parser<camp::list<Tail...> >::checkArgs());
+  }
+};
+
+template<typename List>
+struct call_extractor
+{};
+
+template<typename ...Args>
+struct call_extractor<camp::list<Args...>>
+{
+  template<typename Data>
+  RAJA_HOST_DEVICE
+  static auto make_tuple(Data &&data)
+    -> decltype(camp::make_tuple(extractor<Args>::extract_arg(data) ...))
+  {
+    return camp::make_tuple(extractor<Args>::extract_arg(data) ...);
```

Comment:
You could probably move call_extractor into the invoke_lambda_with_args function.  This is probably an unnecessary class.